---
id: 666832e427d70bc5219dc62a
title: Passo 36
challengeType: 20
dashedName: step-36
---

# --description--

Como você pode ver na saída, embora você tenha definido `__getattr__` na sua classe, esse método ainda não foi chamado.Isso acontece porque o acesso ao atributo padrão ocorre através de `__getattribute__`.Portanto, você pode ver o valor do atributo impresso no terminal.

Agora modifique as duas últimas linhas de código para acessar o atributo `z` de `v1` em ambas as suas chamadas de `print()`.Desta vez, você vai acessar um atributo que `v1` não possui.Como resultado, `__getattribute__` gerará um erro e `__getattr__` será chamado.

# --hints--

Você deve modificar seu código para imprimir o atributo `z` de `v1` usando tanto o operador ponto quanto a função `getattr()`.

```js
({ test: () => assert(runPython(`
n = _Node(_code)
(n.has_call("print(v1.z)") and n.has_call("print(getattr(v1, 'z'))")) or n.has_call("print(v1.z, getattr(v1, 'z'))") or n.has_call("print(getattr(v1, 'z'), v1.z)")
`)) })
```

# --seed--

## --seed-contents--

```py
class R2Vector:
    def __init__(self, *, x, y):
        self.x = x
        self.y = y

    def norm(self):
        return sum(val**2 for val in vars(self).values())**0.5

    def __str__(self):
        return str(tuple(getattr(self, i) for i in vars(self)))

    def __repr__(self):
        arg_list = [f'{key}={val}' for key, val in vars(self).items()]
        args = ', '.join(arg_list)
        return f'{self.__class__.__name__}({args})'

    def __getattr__(self, attr):                
        return 'calling __getattr__'

class R3Vector(R2Vector):
    def __init__(self, *, x, y, z):
        super().__init__(x=x, y=y)
        self.z = z

v1 = R2Vector(x=2, y=3)
v2 = R3Vector(x=2, y=2, z=3)
# print(f'v1 = {v1}', f'\nrepr = {repr(v1)}')
# print(f'v2 = {v2}', f'\nrepr = {repr(v2)}')
--fcc-editable-region--
print(v1.x)
print(getattr(v1, 'x'))
--fcc-editable-region--
```
